Error during make on Cygwin

Hello,
I'm trying to compile the project on Windows 10 with cygwin.

I have cloned the repo and when I execute the line "make". I'm getting the following errors;

> $ make
mkdir -p build/output/samples
mkdir -p build/output/test
echo OSTYPE is Windows_NT
OSTYPE is Windows_NT
cc -g -fPIC  -Os -Wall -fvisibility=hidden -Ibuild  -o build/output/libpaho-mqtt3c.so.1.0 src/Tree.c src/SocketBuffer.c src/utf-8.c src/MQTTReasonCodes.c src/LinkedList.c src/MQTTPersistence.c src/Log.c src/MQTTClient.c src/SHA1.c src/Thread.c src/OsWrapper.c src/Clients.c src/WebSocket.c src/MQTTPacketOut.c src/Messages.c src/MQTTPersistenceDefault.c src/MQTTProtocolClient.c src/MQTTProperties.c src/Base64.c src/StackTrace.c src/MQTTProtocolOut.c src/MQTTPacket.c src/Socket.c src/Heap.c  -shared -Wl,-init, -lpthread
src/SHA1.c: In function ‘SHA1_ProcessBlock’:
src/SHA1.c:109:10: warning: implicit declaration of function ‘be32toh’ [-Wimplicit-function-declaration]
   w[i] = be32toh(w[i]);
          ^~~~~~~
src/SHA1.c: In function ‘SHA1_Final’:
src/SHA1.c:156:15: warning: implicit declaration of function ‘htobe32’ [-Wimplicit-function-declaration]
  ctx->w[14] = htobe32((uint32_t)(total >> 32));
               ^~~~~~~
src/StackTrace.c: In function ‘StackTrace_entry’:
src/StackTrace.c:110:34: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
   Log_stackTrace(trace_level, 9, (int)my_thread->id, my_thread->current_depth, name, line, NULL);
                                  ^
src/StackTrace.c: In function ‘StackTrace_exit’:
src/StackTrace.c:134:36: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
    Log_stackTrace(trace_level, 10, (int)my_thread->id, my_thread->current_depth, name, line, NULL);
                                    ^
src/StackTrace.c:136:36: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
    Log_stackTrace(trace_level, 11, (int)my_thread->id, my_thread->current_depth, name, line, (int*)rc);
                                    ^
/tmp/cceX9sG9.o: In function `SHA1_ProcessBlock':
/home/soulflyer/paho.mqtt.c/src/SHA1.c:109: undefined reference to `be32toh'
/home/soulflyer/paho.mqtt.c/src/SHA1.c:109:(.text+0x48): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `be32toh'
/tmp/cceX9sG9.o: In function `SHA1_Final':
/home/soulflyer/paho.mqtt.c/src/SHA1.c:156: undefined reference to `htobe32'
/home/soulflyer/paho.mqtt.c/src/SHA1.c:156:(.text+0x256): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `htobe32'
/home/soulflyer/paho.mqtt.c/src/SHA1.c:157: undefined reference to `htobe32'
/home/soulflyer/paho.mqtt.c/src/SHA1.c:157:(.text+0x262): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `htobe32'
/home/soulflyer/paho.mqtt.c/src/SHA1.c:162: undefined reference to `htobe32'
/home/soulflyer/paho.mqtt.c/src/SHA1.c:162:(.text+0x275): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `htobe32'
collect2: error: ld returned 1 exit status
make: *** [Makefile:245: build/output/libpaho-mqtt3c.so.1.0] Error 

I'm guessing there is a problem with OpenSSL but I couldn't figure out how to make it work.

when I try " cmake -G "Unix Makefiles" /home/soulflyer/paho.mqtt.c -DPAHO_WITH_SSL=TRUE -DPAHO_BUILD_STATIC=TRUE -DOPENSSL_SEARCH_PATH=/usr/local"

> --CMake version: 3.6.2
-- CMake system name: CYGWIN
-- Timestamp is 2018-10-22T21:18:14Z
-- OpenSSL hints: /usr/local
-- OpenSSL headers found at /usr/local/include
-- OpenSSL library found at /lib/libssl.dll.a
-- OpenSSL Crypto library found at /lib/libcrypto.dll.a
-- OpenSSL hints: /usr/local
-- OpenSSL headers found at /usr/local/include
-- Configuring done
-- Generating done
-- Build files have been written to: /home/soulflyer/paho.mqtt.c/build

And then when I use "make" under build folder

> $ make
[ 22%] Built target common_obj
Scanning dependencies of target paho-mqtt3a
[ 23%] Building C object src/CMakeFiles/paho-mqtt3a.dir/MQTTAsync.c.o
[ 24%] Linking C shared library cygpaho-mqtt3a-1.dll
CMakeFiles/common_obj.dir/SHA1.c.o:SHA1.c:(.text+0xd9): undefined reference to `be32toh'
CMakeFiles/common_obj.dir/SHA1.c.o:SHA1.c:(.text+0xd9): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `be32toh'
CMakeFiles/common_obj.dir/SHA1.c.o:SHA1.c:(.text+0x3d1): undefined reference to `htobe32'
CMakeFiles/common_obj.dir/SHA1.c.o:SHA1.c:(.text+0x3d1): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `htobe32'
CMakeFiles/common_obj.dir/SHA1.c.o:SHA1.c:(.text+0x3e5): undefined reference to `htobe32'
CMakeFiles/common_obj.dir/SHA1.c.o:SHA1.c:(.text+0x3e5): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `htobe32'
CMakeFiles/common_obj.dir/SHA1.c.o:SHA1.c:(.text+0x417): undefined reference to `htobe32'
CMakeFiles/common_obj.dir/SHA1.c.o:SHA1.c:(.text+0x417): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `htobe32'
collect2: error: ld returned 1 exit status
make[2]: *** [src/CMakeFiles/paho-mqtt3a.dir/build.make:139: src/cygpaho-mqtt3a-1.dll] Error 1
make[1]: *** [CMakeFiles/Makefile2:110: src/CMakeFiles/paho-mqtt3a.dir/all] Error 2
make: *** [Makefile:161: all] Error 2


Thanks for your help

--- a/src/MQTTVersion.c
+++ b/src/MQTTVersion.c
@@ -201,7 +201,9 @@ int main(int argc, char** argv)
 		 
 		for (i = 0; i < ARRAY_SIZE(libraries); ++i)
 		{
-#if defined(WIN32) || defined(WIN64)
+#if defined(__CYGWIN__)
+			sprintf(namebuf, "cyg%s-1.dll", libraries[i]);
+#elif defined(WIN32) || defined(WIN64)
 			sprintf(namebuf, "%s.dll", libraries[i]);
 #elif defined(OSX)
 			sprintf(namebuf, "lib%s.1.dylib", libraries[i]);

--- a/src/MQTTVersion.c
+++ b/src/MQTTVersion.c
@@ -138,9 +138,8 @@ int loadandcall(const char* libname)
 	int rc = 0;
 	MQTTAsync_nameValue* (*func_address)(void) = NULL;
 #if defined(WIN32) || defined(WIN64)
-	HMODULE APILibrary;
-
-	if ((APILibrary = LoadLibraryA(libname)) == NULL)
+	HMODULE APILibrary = LoadLibraryA(libname);
+	if (APILibrary == NULL)
 		printf("Error loading library %s, error code %d\n", libname, GetLastError());
 	else
 	{
@@ -153,11 +152,10 @@ int loadandcall(const char* libname)
 	}
 #else
 	void* APILibrary = dlopen(libname, RTLD_LAZY); /* Open the Library in question */
-	char* ErrorOutput = dlerror(); 	               /* Check it opened properly */
-	if (ErrorOutput != NULL)
-		printf("Error loading library %s, error %s\n", libname, ErrorOutput);
+	if (APILibrary == NULL)
+		printf("Error loading library %s, error %s\n", libname, dlerror());
 	else
-	{	
+	{
 		*(void **) (&func_address) = dlsym(APILibrary, "MQTTAsync_getVersionInfo");
 		if (func_address == NULL)
 			func_address = dlsym(APILibrary, "MQTTClient_getVersionInfo");

--- a/src/SHA1.c
+++ b/src/SHA1.c
@@ -53,7 +53,7 @@ int SHA1_Final(unsigned char *md, SHA_CTX *c)
 }
 
 #else /* if defined(WIN32) || defined(WIN64) */
-#if defined(__linux__)
+#if defined(__linux__) || defined(__CYGWIN__)
 #  include <endian.h>
 #elif defined(__APPLE__)
 #  include <libkern/OSByteOrder.h>

